cmake/googletest.cmake

```diff
@@ -0,0 +1,40 @@
+include(FetchContent)
+include(GoogleTest)
+
+set(GOOGLETEST_DIR "" CACHE STRING "Location of local GoogleTest repo to build against")
+
+if(GOOGLETEST_DIR)
+  set(FETCHCONTENT_SOURCE_DIR_GOOGLETEST ${GOOGLETEST_DIR} CACHE STRING "GoogleTest source directory override")
+endif()
+
+message(STATUS "Fetching GoogleTest")
+
+list(APPEND GTEST_CMAKE_CXX_FLAGS 
+     -Wno-undef
+     -Wno-reserved-identifier
+     -Wno-global-constructors
+     -Wno-missing-noreturn
+     -Wno-disabled-macro-expansion
+     -Wno-used-but-marked-unused
+     -Wno-switch-enum
+     -Wno-zero-as-null-pointer-constant
+     -Wno-unused-member-function
+     -Wno-comma
+     -Wno-old-style-cast
+)
+message(STATUS "Suppressing googltest warnings with flags: ${GTEST_CMAKE_CXX_FLAGS}")
+
+FetchContent_Declare(
```

Comment:
Why should we use this function? I don't want to update docker and cmake version just to make a git clone

Reply:
@shurale-nkn , thanks for the comments. If you refer to FetchContent_Declare, this is to declare a dependency on GoogleTest which will be downloaded from GitHub. GoogleTest is a newly integrated framework that's needed for unit testing. 

Only CMake 3.11 and later supports FetchContent. Does this make sense? Thanks

Reply:
@xinlipn 
This does not explain why it cannot be done without this function. And without new cmake.
What does this function give besides the implementation of git clone?
Our dockers already contain git and we used other repositories before GoogleTest .
What's the problem with using GoogleTest without this function?

Reply:
@shurale-nkn 
FetchContent module makes content available immediately, allowing the configure step to use the content in commands like add_subdirectory(), include(), or file() operations. As part of FetchContent, FetchContent_Declare() is not just git clone, it records the options that describe how to populate the specified content. This feature is only supported by CMake 3.11 or later

Without which, CMake would throw errors below:

CMake Error at /usr/share/cmake-3.23/Modules/FetchContent.cmake:754 (message):
  No content details recorded for googletest
Call Stack (most recent call first):
  /usr/share/cmake-3.23/Modules/FetchContent.cmake:1140 (__FetchContent_getSavedDetails)
  /usr/share/cmake-3.23/Modules/FetchContent.cmake:1260 (FetchContent_Populate)
  cmake/googletest.cmake:35 (FetchContent_MakeAvailable)
  test/gtest/CMakeLists.txt:1 (include)

Reply:
not sure why it's better than git submodule.

But ok, as it turned out, we are already using CMake 3.15, so your changes in Dockerfile are redundant.
If the PR will be merged earlier than the fix for issue #1706, no changes are required.

Reply:
@xinlipn
wait, what is written in the error logs?
used a newer version of Cmake`/usr/share/cmake-3.23` and it returned an error at your function? 😄

Reply:
Thanks @shurale-nkn , let me try on the new CMake version, in the meantime, do you have the logs with the erro?

Reply:
@xinlipn 
I'm talking about your log from a previous comment

> @shurale-nkn FetchContent module makes content available immediately, allowing the configure step to use the content in commands like add_subdirectory(), include(), or file() operations. As part of FetchContent, FetchContent_Declare() is not just git clone, it records the options that describe how to populate the specified content, i.e. googletest in this case. This feature, i.e. FetechContent is only supported by CMake 3.11 or later
> 
> Without which, CMake would throw errors below:
> 
> CMake Error at /usr/share/cmake-3.23/Modules/FetchContent.cmake:754 (message): No content details recorded for googletest Call Stack (most recent call first): /usr/share/cmake-3.23/Modules/FetchContent.cmake:1140 (__FetchContent_getSavedDetails) /usr/share/cmake-3.23/Modules/FetchContent.cmake:1260 (FetchContent_Populate) cmake/googletest.cmake:35 (FetchContent_MakeAvailable) test/gtest/CMakeLists.txt:1 (include)
> 
> Not sure if this helps. Thanks

Reply:
> @xinlipn wait, what is written in the error logs? used a newer version of Cmake`/usr/share/cmake-3.23` and it returned an error at your function? 😄

@shurale-nkn , I see. This CMake error was introduced after removing the FetchContent to address your comments earlier "What's the problem with using GoogleTest without this function?"

There's no CMake error with FetchContent :)

Reply:
oh..., well, then it seems obvious that the second function with same prefix will not work and it will be necessary to assemble the package differently. no one canceled the build setup.

Reply:
> oh..., well, then it seems obvious that the second function with same prefix will not work and it will be necessary to assemble the package differently. no one canceled the build setup.

Sorry if I wasn't clear.  The CMake error listed earlier was caused by removing FetchContent_Declare(). Not sure I totally get what you mean by "no one canceled the build setup". Could you elaborate?

Reply:
I meant that if we replaced the method of receiving the package, we need to replace the function responsible for its assembly. And for the assembly, it may be necessary to set the correct settings of the environment variables. 
But in this case, you will be sure of what you are doing and how the program should behave.

The "FetchContent" is not the only solution for dependency control.

Reply:
@shurale-nkn  Thanks for the comment, let me fix this via requirements.txt in a followup PR, tracked in issue #1838